Remove older style variables and replaced with newer ones

clarkepaul opened the pull request:
Replaced older variables with newer ones where possible.
Most notable visual differences are seen in the preview/edit mode toggle dropdown:
![image](https://user-images.githubusercontent.com/555033/32488482-9b969650-c411-11e7-8639-3aacb0274978.png)
 
Requires asset-admin PR https://github.com/silverstripe/silverstripe-asset-admin/pull/690
Related/contributes to issue https://github.com/silverstripe/silverstripe-admin/issues/201

I've checked the main repo's (security, cms, framework, campaign-admin) to make sure they weren't being referenced but it could be that some might be used in modules? CWP modules are being re-checked with SS4 in the coming months so it would be good to see this get in before all of the main modules get re-checked with SS4.

flamerohr commented on client/src/styles/_bootstrap-variables.scss:
```diff
@@ -50,16 +50,15 @@
+$gray-500: lighten($gray-700, 20%) !default;
+$gray-600: lighten($gray-700, 10%) !default;
+$gray-800: darken($gray-700, 10%) !default;
+$gray-900: darken($gray-700, 20%) !default;
```
I don't mind using `lighten()` and `darken()` in variables - could we put the calculated colour codes next to them in comment?
Makes searching for colour codes much much easier e.g. especially when we get design details like `#ced4da` and we have no idea which one that is... :/

clarkepaul replied:
I could add them. I am setting up an easier way to get all the colours through this tool which syncs with the design doc and should match the variables sheet also https://brand.ai/silverstripe/cms-library

It will all be clearer with the pattern library too—when we have it.

phalkunz replied:
@flamerohr Are you saying you should put a color code in a comment where there's color calculation or just this section? I can see it being really helpful for searching and debugging purposes but it'd be a pain to have to do that everywhere.

clarkepaul replied:
Probably just on these grays it would be okay.

flamerohr replied:
just on these greys would already be a big help :)

flamerohr commented on client/src/styles/_bootstrap-variables.scss:
```diff
@@ -79,12 +77,12 @@ $blue: #005a93; // custom
 $purple:  #6f42c1 !default; // bootstrap default
 $pink:    #e83e8c !default; // bootstrap default
-$red: #d40404; // custom
+$red:     #DA273B !default; // custom
```
mix of upper case colours and lower case colours, could we stick to one for consistency? ;)

clarkepaul replied:
Yup easy fix, typically all uppercase is cleaner.

phalkunz replied:
Also the spacing between variables and their values.

clarkepaul replied:
Seems Bootstrap use all lower so best to stick to that actually :)

flamerohr commented on client/src/styles/_bootstrap-variables.scss:
```diff
@@ -165,7 +159,7 @@ $spacers: (
 // Settings for the `<body>` element.
 
-$body-bg: #f6f7f8;
+$body-bg: lighten($gray-700, 63);
```
is this different from the other `gray-*` colour codes? are we able to use those instead of another variation?
(p.s. I don't mind if this is kept, but could we have the computed colour code commented here as well?)

flamerohr commented on client/src/styles/_bootstrap-variables.scss:
```diff
@@ -235,14 +229,14 @@ $line-height-lg: (4 / 3);
 
 //$border-width: 1px !default;
-$border-color: darken(#f6f7f8, 10);
+$border-color: lighten($gray-200, 2);
```
same as above, is this different from the other `gray-*` colour codes?

phalkunz replied:
I can see it make sense do this on a color that is used on a large surface but with color to be used with border properly 2% is hardly any different.

flamerohr commented on client/src/styles/_typography.scss:
```diff
@@ -86,7 +86,7 @@ th {
   font-size: $font-size-xxs;
   font-weight: 400;
-  letter-spacing: .4px;
+  letter-spacing: .1px;
```
`.1px` is really small, and you could end up with rounding differences across browsers with such a small value.

I believe Firefox rounds UP for this and all other browsers rounds DOWN.

phalkunz replied:
I just tested this out. It quite noticeable between `0px`, `.1px` and `1px` on Chrome and Firefox.

clarkepaul replied:
I think letter spacing is rounding is fine for all browsers. Not sure why I reduced that as .4px looked fine actually (probably .3 looks even better).

clarkepaul replied:
Going with .3px. Left is Chrome and right is Firefox.
![image](https://user-images.githubusercontent.com/555033/32870612-9e7bd78c-cae1-11e7-9f20-b656a0b63cbc.png)

flamerohr replied:
sweet :)

flamerohr commented on client/src/styles/_variables.scss:
```diff
@@ -147,10 +144,11 @@ $z-depth-2: 0 8px 12px 0 $shadow-level-3, 0 6px 20px 0 $shadow-level-2;
+// Menu theme colours
 
+$base-menu: #d8e4eb !default;
+$base-menu-bg: 	lighten($base-menu, 5%);
```
spacing is a bit wierd here

clarkepaul replied:
Cool can fix

phalkunz commented on client/src/styles/_variables.scss:
```diff
@@ -9,31 +9,29 @@ $brand-primary: #29abe2;
-$border-color-light: darken($body-bg, 6);
-$border-color-dark: darken($body-bg, 14);
+$border-color-light: lighten($border-color, 4);
+$border-color-dark: darken($border-color, 2);
```
The same above comment about border colors.

clarkepaul commented:
These variables are in the legacy folder already (from 2.4!) and shouldn't really be used but they are continued to be used and updated. We currently have three variables files which are already super confusing, maybe we could leave the deleted ones and put a note not to use them so that we can delete later. The main thing I was trying to achieve is to get everything using the right variables as the code is so misaligned with designs at the moment, so not a biggy leaving the old ones there.

clarkepaul commented:
@flamerohr @phalkunz I think I got all recommendations and I've added back all the unused variables :( I don't have time to go through them all and link/reference the newer ones in case they do get used—there are many which don't have a newer similar variable. I still think this is making good progress as it stands. I did let @chillu know I wanted to get rid of all unused variables and I think from memory it wasn't a deal breaker but shouldn't matter anymore.

flamerohr commented:
@clarkepaul sweet, looking good :)

hmm... I guess I'd prefer it was a warning if used than outright removed... where the warning would be butt-ugly but still work. (could make them all Bright green colours and giant width/height values)

Do you know if the CC's will be importing the legacy variables (assuming we put it in a separate file)? I guess I'd ask if they could exclude that - so then the unused variables wouldn't work for them.

I can investigate if there's a better way to do this in scss

clarkepaul commented:
I talked to @robbieaverill who possibly had a good idea which was to lint the older variables so we get warnings when they get used but thought it might be a fair bit of work.

All the modules would be importing the legacy variables as far as I'm aware.

Let me know if you think it needs any more changes.

chillu commented:
As stated in other contexts, I don't think rigorously applying sermver on "frontend APIs" like SCSS variable names will do us much good in the 4.x release line. We're just not stable on this stuff yet, and need to move relatively fast in terms of getting the ReactJS and Bootstrap transition over the line. So I don't think we should keep unused SCSS variables around in core just in case someone built their own SCSS on top of it. We'll want more stability around those around a 4.3 release or even a 5.0, but until then, building CSS functionality on top of the CMS will always require more upgrading efforts.

That being said, if there's a low impact way to mark things like these as deprecated without impacting the overall CMS frontend developer experience, sure thing.

clarkepaul commented:
Rebased and removed bug of square corners 
![pasted_image_21_11_17__12_03_pm](https://user-images.githubusercontent.com/555033/33046030-1bf8cf50-ceb4-11e7-8413-5485a8c8f560.png)

flamerohr commented:
Sweet as, retaining variables were the only contentious point I had left.

Everything else has been addressed, and great effort from @clarkepaul 

So we've shifted from "Gradient EVERYTHING" to "Round ALL THE CORNERS", such a fun world to be in :)
